api/src/main/java/org/xnio/channels/Channels.java

```diff
@@ -311,13 +311,22 @@ public static void shutdownWritesBlocking(SuspendableWriteChannel channel) throw
      * @since 1.2
      */
     public static <C extends ReadableByteChannel & SuspendableReadChannel> int readBlocking(C channel, ByteBuffer buffer, long time, TimeUnit unit) throws IOException {
+        // In the fast path, the timeout is not used because bytes can be read without blocking.
         int res = channel.read(buffer);
-        if (res == 0 && buffer.hasRemaining()) {
-            channel.awaitReadable(time, unit);
-            return channel.read(buffer);
-        } else {
+        if (res != 0) {
             return res;
         }
+        long startTimeNanos = System.nanoTime();
+        long timeoutNanos = unit.toNanos(time);
+        long remainingNanos;
+        while (res == 0 && buffer.hasRemaining()
+                // Note: nanotime must only be used in comparison with another nanotime value
+                && (remainingNanos = timeoutNanos - (System.nanoTime() - startTimeNanos)) > 0) {
```

Comment:
This structure causes `System.nanoTime` to be called twice in short succession with no intervening wait.  It's preferred to call `nanoTime` once at the start, then immediately begin waiting.